File: scaly/src/containers/hashmap.rs
use containers::array::Array;
use containers::hashset::Hash;
use containers::hashset::HashPrimeHelper;
use containers::list::List;
use containers::reference::Ref;
use containers::vector::Vector;
use memory::page::Page;
use memory::region::Region;

#[derive(Copy, Clone)]
pub struct KeyValuePair<K: Hash<K> + Copy, V: Copy> {
    key: K,
    value: V,
}

#[derive(Copy, Clone)]
pub struct HashMap<K: Hash<K> + Copy, V: Copy> {
    slots: Ref<Vector<Ref<List<Slot<KeyValuePair<K, V>>>>>>,
}

impl<K: Hash<K> + Copy, V: Copy> HashMap<K, V> {
    pub fn from_vector(
        _pr: &Region,
        _rp: *mut Page,
        vector: Ref<Vector<KeyValuePair<K, V>>>,
    ) -> Ref<HashMap<K, V>> {
        let _r = Region::create(_pr);
        let hash_size = HashPrimeHelper::get_prime(vector.length);
        let mut array: Ref<Array<Ref<List<Slot<KeyValuePair<K, V>>>>>> =
            Ref::new(_r.page, Array::new());
        for _ in 0..hash_size {
            array.add(Ref::new(_r.page, List::new()));
        }
        let slots = Ref::new(_rp, Vector::from_array(_rp, array));
        let mut hash_map = Ref::new(_rp, HashMap { slots: slots });
        hash_map.initialize_from_vector(vector);
        hash_map
    }

    fn initialize_from_vector(&mut self, vector: Ref<Vector<KeyValuePair<K, V>>>) {
        for key_value_pair in vector.iter() {
            self.add(&key_value_pair.key, &key_value_pair.value);
        }
    }

    fn add(&mut self, key: &K, value: &V) {
        let hash_code = key.hash();
        let slot_number = hash_code % self.slots.length;
        let mut slot_list = self.slots[slot_number];
        for slot in slot_list.get_iterator() {
            if key.equals(&slot.value.key) {
                return;
            }
        }

        slot_list.add(Slot {
            hash_code: hash_code,
            value: KeyValuePair {
                key: *key,
                value: *value,
            },
        });
    }

    pub fn contains(&self, value: K) -> bool {
        for slot in self.slots[value.hash() % self.slots.length].get_iterator() {
            if value.equals(&slot.value.key) {
                return true;
            }
        }

        false
    }
}

#[derive(Copy, Clone)]
struct Slot<K: Copy> {
    value: K,
    hash_code: usize,
}

#[test]
fn test_hash_map() {
    use containers::String;
    use memory::Heap;
    use memory::Page;
    use memory::Region;
    use memory::StackBucket;
    let mut heap = Heap::create();
    let root_stack_bucket = StackBucket::create(&mut heap);
    let root_page = Page::get(root_stack_bucket as usize);
    let _r = Region::create_from_page(root_page);
    let keywords = HashMap::from_vector(
        &_r,
        _r.page,
        Ref::new(
            _r.page,
            Vector::from_raw_array(
                _r.page,
                &[
                    KeyValuePair {
                        key: String::from_string_slice(_r.page, "using"),
                        value: 1,
                    },
                    KeyValuePair {
                        key: String::from_string_slice(_r.page, "namespace"),
                        value: 2,
                    },
                    KeyValuePair {
                        key: String::from_string_slice(_r.page, "typedef"),
                        value: 3,
                    },
                ],
            ),
        ),
    );

    assert_eq!(
        (*keywords).contains(String::from_string_slice(_r.page, "using")),
        true
    );
    assert_eq!(
        (*keywords).contains(String::from_string_slice(_r.page, "namespace")),
        true
    );
    assert_eq!(
        (*keywords).contains(String::from_string_slice(_r.page, "typedef")),
        true
    );
    assert_eq!(
        (*keywords).contains(String::from_string_slice(_r.page, "nix")),
        false
    );
}

File: scaly/containers/HashMap.cpp
namespace scaly {
namespace containers {

using namespace scaly::memory;

template<class K, class V> struct HashMap;

template<class K, class V>
struct HashMapIterator : Object {
    Vector<Vector<KeyValuePair<K, V>>>* slots;
    VectorIterator<Vector<KeyValuePair<K, V>>>* slot_iterator;
    VectorIterator<KeyValuePair<K, V>>* element_iterator;

    HashMapIterator<K, V>(const HashMap<K, V>& hash_map) {
        if (hash_map.slots == nullptr)
            return;

        slot_iterator = new (alignof(VectorIterator<Vector<KeyValuePair<K, V>>>), Page::get(this)) VectorIterator<Vector<KeyValuePair<K, V>>>(*hash_map.slots);
        element_iterator = nullptr;
    }

    V* next() {
        while (true)
        {
            if (element_iterator == nullptr) {
                if (slot_iterator == nullptr)
                    return nullptr;
                
                auto slot = slot_iterator->next();
                if (slot == nullptr) {
                    slot_iterator = nullptr;
                    return nullptr;
                }

                element_iterator = new (alignof(VectorIterator<KeyValuePair<K, V>>), Page::get(this)) VectorIterator<KeyValuePair<K, V>>(*slot);
            }

            auto element = element_iterator->next();
            if (element != nullptr)
                return &element->value;
            
            element_iterator = nullptr;
        }
    }
};

template<class K, class V>
struct HashMap : Object {
    Vector<Vector<KeyValuePair<K, V>>>* slots;

    HashMap<K, V>(Page* _rp, HashMapBuilder<K, V>& hash_map_builder) {
        Region _r;
        if (hash_map_builder.length == 0) {
            this->slots = nullptr;
            return;
        }

        this->slots = new(alignof(Vector<Vector<KeyValuePair<K, V>>>), _rp) Vector<Vector<KeyValuePair<K, V>>>(_rp, hash_map_builder.slots->length);
        auto length = hash_map_builder.slots->length;
        for (size_t i = 0; i < length; i++) {
            Region _r_1;
            Array<KeyValuePair<K, V>>& array = *new(alignof(Array<KeyValuePair<K, V>>), _r_1.get_page()) Array<KeyValuePair<K, V>>();
            auto list_iterator = ListIterator<Slot<KeyValuePair<K, V>>>(hash_map_builder.slots->get(i)->head);
            while (auto item = list_iterator.next())
                array.add(item->value);
            if (array.length > 0)
                this->slots->set(i, Vector<KeyValuePair<K, V>>(_rp, array));
        }
    }

    bool contains(const K& key) {
        if (this->slots == nullptr)
            return false;

        auto hash = key.hash();
        auto slot_number = hash % this->slots->length;
        auto slot = this->slots->get(slot_number);
        auto length = slot->length;
        for (size_t i = 0; i < length; i++) {
            if (key.equals(slot->get(i)->key)) {
                return true;
            }
        }

        return false;
    }

    V* operator [](const K& key) {
        if (this->slots == nullptr)
            return nullptr;

        auto hash = key.hash();
        auto slot_number = hash % this->slots->length;
        auto slot = this->slots->get(slot_number);
        auto length = slot->length;
        for (size_t i = 0; i < length; i++) {
            if (key.equals(slot->get(i)->key)) {
                return &(slot->get(i)->value);
            }
        }

        return nullptr;
    }

    Vector<V*>& get_values(Page* _rp) const {
        Region _r;
        auto array = *new(alignof(Array<V*>), _r.get_page()) Array<V*>();
        auto slot_iterator = VectorIterator<Vector<KeyValuePair<K, V>>>(*this->slots);
        while (auto slot = slot_iterator.next()) {
            auto element_iterator = VectorIterator<KeyValuePair<K, V>>(*slot);
            while (auto element = element_iterator.next()) {
                array.add(&element->value);
            }
        }

        return *new(alignof(Vector<V*>), _rp) Vector<V*>(_rp, array);
    }
};

}
}